internal/topology/variables/cluster_variable_validation_test.go

```diff
@@ -1899,6 +1899,447 @@ func Test_ValidateClusterVariable(t *testing.T) {
+				},
+			},
+			wantErrs: []validationMatch{
+				invalidType(`allOfExampleField in body must be of type integer: "string"`, "spec.topology.variables[test].value.allOfExampleField"),
+				invalid(
+					`Invalid value: "{\"anyOfExampleField\":\"42\", \"allOfExampleField\":\"42\", \"oneOfExampleField\":\"42\"}": "allOfExampleField" must validate all the schemas (allOf)`,
+					"spec.topology.variables[test].value",
+				),
+			},
+		}, {
+			name: "fail for int-or-string in oneOf/anyOf/allOf schemas with object values",
+			clusterClassVariable: &clusterv1.ClusterClassVariable{
+				Name: "test",
+				Schema: clusterv1.VariableSchema{
+					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+						Type: "object",
+						Properties: map[string]clusterv1.JSONSchemaProps{
+							"anyOfExampleField": { // Valid variant
+								AnyOf: []clusterv1.JSONSchemaProps{{
+									Type: "integer",
+								}, {
+									Type: "string",
+								}},
+							},
+							"allOfExampleField": { // Invalid variant
+								AllOf: []clusterv1.JSONSchemaProps{{
+									Type: "integer",
+								}, {
+									Type: "string",
+								}},
+							},
+							"oneOfExampleField": { // Valid variant
+								OneOf: []clusterv1.JSONSchemaProps{{
+									Type: "integer",
+								}, {
+									Type: "string",
+								}},
+							},
+						},
+					},
+				},
+			},
+			clusterVariable: &clusterv1.ClusterVariable{
+				Name: "test",
+				Value: apiextensionsv1.JSON{
+					Raw: []byte(`{"anyOfExampleField":{}, "allOfExampleField":{}, "oneOfExampleField":{}}`),
+				},
+			},
+			wantErrs: []validationMatch{
+				invalid(
+					`Invalid value: "{\"anyOfExampleField\":{}, \"allOfExampleField\":{}, \"oneOfExampleField\":{}}": "anyOfExampleField" must validate at least one schema (anyOf)`,
+					"spec.topology.variables[test].value",
+				),
+				invalidType(`anyOfExampleField in body must be of type integer: "object"`, "spec.topology.variables[test].value.anyOfExampleField"),
+				invalidType(`allOfExampleField in body must be of type integer: "object"`, "spec.topology.variables[test].value.allOfExampleField"),
+				invalidType(`allOfExampleField in body must be of type string: "object"`, "spec.topology.variables[test].value.allOfExampleField"),
+				invalid(
+					`Invalid value: "{\"anyOfExampleField\":{}, \"allOfExampleField\":{}, \"oneOfExampleField\":{}}": "allOfExampleField" must validate all the schemas (allOf). None validated`,
+					"spec.topology.variables[test].value",
+				),
+				invalid(
+					`Invalid value: "{\"anyOfExampleField\":{}, \"allOfExampleField\":{}, \"oneOfExampleField\":{}}": "oneOfExampleField" must validate one and only one schema (oneOf). Found none valid`,
+					"spec.topology.variables[test].value",
+				),
+				invalidType(`oneOfExampleField in body must be of type integer: "object"`, "spec.topology.variables[test].value.oneOfExampleField"),
+			},
+		},
```

Comment:
Sorry this comment contained the wrong test case: https://github.com/kubernetes-sigs/cluster-api/pull/10637#discussion_r1776732476 :)

I was suggesting to add this test case here:

```go
		{
			name: "Valid object for int-or-string (resource.Quantity)",
			clusterClassVariable: &clusterv1.ClusterClassVariable{
				Name:     "quantityArray",
				Required: true,
				Schema: clusterv1.VariableSchema{
					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
						Type: "array",
						Items: &clusterv1.JSONSchemaProps{
							AnyOf: []clusterv1.JSONSchemaProps{{
								Type: "integer",
							}, {
								Type: "string",
							}},
							Pattern: `^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$`,
						},
					},
				},
			},
			clusterVariable: &clusterv1.ClusterVariable{
				Name: "quantity",
				Value: apiextensionsv1.JSON{
					Raw: []byte(`["500000G", "200M", "0.2G"]`),
				},
			},
		},

Reply:
Adding test case but moving pattern next to string typed optional schema and adding integer value to array to test integer optional schema.

Reply:
Asnoted above. This only actually makes sense / works if we have int-or-string in the schema

Reply:
@jimmidyson ^^ I think either way without XIntOrString it won't be valid

Upstream field:
```
	XIntOrString bool `json:"x-kubernetes-int-or-string,omitempty" protobuf:"bytes,40,opt,name=xKubernetesIntOrString"`
```

Reply:
Moved pattern back outside to be consistent with `resource.Quantity` schema.

Reply:
Makes sense, thx!

Reply:
Updated in commit that supports XIntOrString so it is a valid schema.